### PR TITLE
[Relay/Topi][Op] Added native DepthToSpace and SpaceToDepth Operators

### PR DESCRIPTION
**Pull request by jwfromm**
Although we previously supported DepthToSpace and SpaceToDepth in our frontends through the stacking of reshape, transpose, and another reshape operators, it was a clumsy and slow method as pointed out in Issue #4523. This PR adds direct DepthToSpace and SpaceToDepth operators in TOPI and corresponding Relay operations. I also replace the existing frontend calls to DepthToSpace and SpaceToDepth to show the simplification. In my brief testing, these new operators are around 1.5 to 2 times faster than our previous approach.

### CODE REVIEW

**Comment by jwfromm** [comment]
@icemelon9, can you take a look at this PR?

File: python/tvm/relay/frontend/onnx.py [hunk 1]
```diff
@@ -539,35 +539,7 @@ class DepthToSpace(OnnxOpConverter):
...
-            # expand input to larger dimension.
-            expanded = _op.reshape(inputs[0],
-                                   newshape=(in_n, new_c, block_size, block_size, in_h, in_w))
-            # reorder to expand spatial blocks.
-            transposed = _op.transpose(expanded, axes=(0, 1, 4, 2, 5, 3))
-
-        return AttrCvt(op_name="reshape",
-                       extras={'newshape': newshape},
-                       ignores=['mode', 'blocksize'])([transposed], attr)
+        return _op.nn.depth_to_space(inputs[0], block_size)
```
Comment (kice) [inline python/tvm/relay/frontend/onnx.py]: We need to handle eacho DepthToSpace mode differently. 

For the record, PixelShuffle in Pytorch (onnx CRD mode) and depth_to_space in tensorflow (onnx DCR mode) is different.

Comment (jwfromm) [reply python/tvm/relay/frontend/onnx.py]: Thank you for the correction. I never quite understood the use of CDR mode, but now that I know its how PyTorch handles depthtospace it makes much more sense. I added a mode argument to the depth_to_space operator that allows switching the channel indexing. Also added the proper mode parsing in the ones frontend.

**Comment by tqchen** [comment]
@kice please https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

**Comment by tqchen** [comment]
Thanks @jwfromm @kice !

**Comment by CoinCheung** [comment]
Hi, Does tvm fully support the op of depth_to_space now ? 
I got the error message in my code that I do not know whether it is associated with this operator: 
```
Traceback (most recent call last):

  File "run_onnx.py", line 88, in <module>
    graph, lib, params = compile_model()

  File "run_onnx.py", line 53, in compile_model
    mod, params = relay.frontend.from_onnx(model, shape_dict, dtype='float32', opset=11)

  File "/root/build/tvm/python/tvm/relay/frontend/onnx.py", line 1586, in from_onnx
    mod, params = g.from_onnx(graph, opset)

  File "/root/build/tvm/python/tvm/relay/frontend/onnx.py", line 1414, in from_onnx
    op = self._convert_operator(op_name, inputs, attr, opset)

  File "/root/build/tvm/python/tvm/relay/frontend/onnx.py", line 1514, in _convert_operator
    sym = convert_map[op_name](inputs, attrs, self._params)

  File "/root/build/tvm/python/tvm/relay/frontend/onnx.py", line 574, in _impl_v11
    return _op.nn.depth_to_space(inputs[0], block_size, mode=mode)

  File "/root/build/tvm/python/tvm/relay/op/nn/nn.py", line 2374, in depth_to_space
    return _make.depth_to_space(data, block_size, layout, mode)

  File "tvm/_ffi/_cython/./function.pxi", line 304, in tvm._ffi._cy3.core.FunctionBase.__call__

  File "tvm/_ffi/_cython/./function.pxi", line 248, in tvm._ffi._cy3.core.FuncCall

  File "tvm/_ffi/_cython/./function.pxi", line 172, in tvm._ffi._cy3.core.make_arg

TypeError: Don't know how to handle type <class 'bytes'>
```

**Comment by kice** [comment]
Yes it does, but this pr forgot to decode the byte to string. You could add a line to decode the `mode` to string.

**Comment by CoinCheung** [comment]
@kice Thanks for replying !!! I am not familiar with the source code, would you please add a pr to fix this, or would you tell me where should I modify to make my code work ?

**Comment by kice** [comment]
Please give it a try https://github.com/apache/incubator-tvm/pull/4750

**Comment by CoinCheung** [comment]
Thanks, it works for me.

**Pull request merged**